src/spec-lang/tc/typecheck.ts

```diff
@@ -1500,6 +1514,66 @@ function matchArguments(
     return true;
 }
 
+/**
+ * We check the following in forall (uint t in [a ... b]) e(t):
+ *   - Type of t should be numeric
+ *   - a and b should be numeric and be castable to the type of t
+ *   - type of e is boolean
+ *   - return value of this expression is boolean
+ *   - t is defined in e(t).
+ */
+export function tcForAll(expr: SForAll, ctx: STypingCtx, typeEnv: TypeEnv): TypeNode {
+    assert(
+        typeof expr.start == typeof expr.end,
+        `The types of ${expr.start} and ${expr.end} are unequal, One of them is likely undefined`
+    );
+    const startT = tc(expr.start(), ctx, typeEnv);
+
+    if (!(startT instanceof IntType || startT instanceof IntLiteralType)) {
+        throw new SWrongType(
+            `The expected type for ${expr.start().pp()} is numeric and not ${startT}.`,
+            expr.start(),
+            startT
+        );
+    }
+
+    const endT = tc(expr.end(), ctx, typeEnv);
+    if (!(startT instanceof IntType || startT instanceof IntLiteralType)) {
+        throw new SWrongType(
+            `The expected type for ${expr.end().pp()} is numeric and not ${endT}.`,
+            expr.end(),
+            endT
+        );
+    }
+
+    if (!isImplicitlyCastable(expr, startT, expr.iteratorType)) {
+        throw new SWrongType(
+            `The type for ${expr.start().pp()} is not castable to ${expr.iteratorType}.`,
+            expr.start(),
+            expr.iteratorType
+        );
+    }
+
+    if (!isImplicitlyCastable(expr, endT, expr.iteratorType)) {
+        throw new SWrongType(
+            `The type for ${expr.end().pp()} is not castable to ${expr.iteratorType}.`,
+            expr.end(),
+            expr.iteratorType
```

Comment:
This should be `endT`. `SWrongType` expects the actual type as the last argument.